### ISSUE TITLE
Add colorpicker component

### ISSUE THREAD

**Issue by vargasd** [salesforce/design-system-react#1321]
Looking to add a component for the [SLDS colorpicker](https://www.lightningdesignsystem.com/components/color-picker/).

Initially, I'm just looking to add support for the base configuration, with plans to add the custom-only, predefined-only, and swatch-only versions later. Props would include:

* assistiveText
* color
* id
* labels
* onChange
* swatchColors

One thing that is not covered in the SLDS colorpicker spec is the idea of having `transparent` as an option. This is needed for our implementation, and I would like to add custom CSS to represent a transparent swatch state. Thoughts?

**Comment by interactivellama** [comment]
Awesome! I know you all are not the only team interested in this component. I'll try to think of any gotchas in the next day or two, and I'll get back to you. This is basically an `input` with a dropdown so you'll probably want `value` and some other `input` props

@zahnster worked on the SLDS implementation. Jade, do you have any thoughts on a transparent swatch feature?

**Comment by interactivellama** [comment]
## Color Picker API - Initial thoughts

### Behavior

I think this may be the most complicated SLDS component UI I've seen so far. It looks more complicated than Datepicker to me—although without the Date parsing. I'm trying to track down a user testing prototype that was made before it was added to SLDS. We may be able to grab some logic from it.

The most important thing I'm trying to consider is whether the menu has it's own state and is uncontrolled. At this point, I'm thinking it should have internal state and just return the value on close/done, since the “current color” does not change until then, and the “working color” is thrown out on cancel. Also, many devs would be annoyed if they had to have a state container/redux for the primary input AND the color pick menus (swatch AND custom). This would limit the number of prop event callbacks, too. Let's pursue uncontrolled with internal state, although we will need initial state passed down into the menu on render, since the color can change via the primary input.

Within the menu: the inputs, swatches, custom dragger should be controlled and have a central top level internal state, so that all the UI components can be as dumb and stateless as possible. I would recommend many sub components for this component in order to keep the logic within each component simple.

Check out Datepicker's calendar for good logic on keyboard events for swatch focus.

Since are so many events and probably more in the future, I would recommend an events object to group them.

### Recommended Props

* assistiveText
    * label - alternative for hidden label
* className - slds-color-picker node
* classNameMenu - slds-popover node
* disabled (optional, since it's not in the SLDS spec, but would prevent menu open and text input)
* errorText
    * for outer input. has error messaging if present. Do not trigger the error internally, but with a state container based on isColorValid being false.
* errorTextMenu
    * for inner input. has error messaging if present within the menu
* hasStaticAlignment - menu position locking
* id - HTML id
* isOpen - boolean for a controlled UI (See Datepicker)
* label - the heading
* labels
    * default
    * custom
    * cancel
    * done
    * r
    * g
    * b
    * hex
* menuPosition
    * all menus have this, just passed down to Dialog
* events.onCancel
    * callback with parameters: event, { trigger: 'cancel' || 'clickOutside' } - This aligns with popover and can be the same callback 
* events.onClose - fired after close (See Datepicker)
* events.onOpen - Fired after open (See Datepicker)
* events.onRequestClose - used with isOpen to create a controlled UI (See Datepicker)
* events.onRequestOpen - used with isOpen to create a controlled UI (See Datepicker)
* events.onChange 
    * input is controlled, manual typing goes through state container, event of input change
    * triggered by done pressed
    * Data Object
        * value - hex value for input 
        * isColorValid - Calls parseColor and return false if it is not a valid hex. This can trigger the error text in the state container. 
* parseColor - a callback function that receives a string from input that can be overridden
    * a default function is needed, so that it is optional. only hex color values are valid
    * return false if invalid
* parseWorkingColor - callback that if it returns false, errorTextMenu is shown
    * Callback parameters: event, { hex, r, g, b }
    * return false if invalid
    * a default function is needed, so that it is optional
* swatchColors
    * array of hex value strings.
    * a default is needed.
* value
    * the primary input value
    * string of hex value with '#' such as '#333'
    * Should “transparent” be a string or undefined? We need that red line bar swatch for transparent. I don't know if SLDS has that. We will also need a way to select transparent and communicate that choice to assistive tech users.

**Comment by vargasd** [comment]
Thanks for the suggestions. I've got a [WIP implementation](https://github.com/vargasd/design-system-react/tree/colorpicker) and most of these are covered, although I have more to expose around the dialog control and consumer color parsing, but it should all be pretty straightforward. And I totally overlooked the swatch interactions and only did the HSV interactions.

I think the "transparent" value might make sense as any falsey value. Thinking about how we'd used this (generally for things like specifying background-color, it's usually a default state and not explicitly set, although I could see uses case for both an unspecified state and a transparent state.

One thing I noticed is that the `parseWorkingColor` handler suggested wouldn't handle highlighting the error states of the inputs. Maybe if it finds invalid states it could return an object with the message and the fields that have errors?

Also, would `validateColor` and `validateWorkingColor` make more sense? There may already be a pattern for using `parseX` to return a boolean, but to me it seems like it would return a new representation of the thing being parsed.

**Comment by interactivellama** [comment]
@BrunoFonzi ColorPicker discussion

**Comment by interactivellama** [comment]
@vargasd What is the status of this component? Do I need to clarify anything? Is it ready for a formal pull request?

**Comment by vargasd** [comment]
@interactivellama I think I've got everything I need. Just need to get some time to sit down and finish implementing everything. I'll aim for this weekend.

**Comment by interactivellama** [comment]
@vargasd Some other scrum teams are asking for Colorpicker and have offered to help with the pull request if needed.

**Comment by vargasd** [comment]
Yeah, that might be helpful. Sorry, this was deprioritized by my team and has completely fallen off my radar.

**Comment by interactivellama** [comment]
@vargasd Could you make a quick WIP list of things you were still wanting to implement, so Karan and I won't have to go searching for feature gaps?

**Comment by vargasd** [comment]
Sure. I believe these are what I have left:

- Validation/custom color handling. All of this is handled internal to the component and needs to be opened up to consumers.
- Swatch keyboard interaction
- Snapshot tests. I've written some mocha tests around most of the user interactions that currently exist, but haven't written any snapshot tests yet.

**Comment by interactivellama** [comment]
@karanb0 Sam says it's mostly done and the above is left.

**Comment by interactivellama** [comment]
@vargasd You've thought about this some. What do you think of allowing CSS Color Module Level 4 values into the ColorPicker? I'm thinking https://caniuse.com/#feat=css-rrggbbaa and so `value="#ffffff00"` would trigger the little red slash for transparent, but still fit the hex value format instead of a falsey value?

Logic: if 9 digit string and last two are `00`, add slash in swatch options?

**Comment by vargasd** [comment]
What about something like a checkered background by default on the swatches? Then you can truly see the opacity level. Otherwise we only support showing the true color if it's fully transparent or fully opaque.

**Comment by interactivellama** [comment]
@karanb0 How is the Colorpicker coming along? Is there anything I can do to help? Do you have an estimated time it could be added.
